### PR TITLE
feat(op-challenger): Claim Loader

### PR DESCRIPTION
**Pull request by refcell**
**Description**

Creates a net-new component to the game-solver/challenger to load and track
`Claim`s from a fault dispute game contract through polling.

**Metadata**

Fixes CLI-4126

### CODE REVIEW

**Comment by refcell** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #6186** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6186" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6186?utm_source=stack-comment).

File: op-challenger/fault/loader.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+		return err
+	}
+
+	// Grab each claim and push it to the [gameState].
+	for i := uint64(0); i < claimCount.Uint64(); i++ {
+		claim, err := l.fetchClaim(ctx, i)
+		if err != nil {
+			return err
+		}
+		if err := l.state.Put(*claim); err != nil {
```
Comment (trianglesphere) [inline op-challenger/fault/loader.go:85]: `FetchClaims` should return the list of claims & then you should insert the claims into the state separately.

File: op-challenger/fault/loader.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	if err != nil {
+		return []Claim{}, err
+	}
+
+	// Fetch each claim and build a list.
+	claimList := make([]Claim, claimCount.Uint64())
+	for i := uint64(0); i < claimCount.Uint64(); i++ {
+		claim, err := l.fetchClaim(ctx, i)
+		if err != nil {
+			return []Claim{}, err
```
Comment (ajsutton) [inline op-challenger/fault/loader.go:84]: nit:
```suggestion
			return nil, err
```

File: op-challenger/fault/loader.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+	return &claim, nil
+}
+
+// FetchClaims fetches all claims from the fault dispute game.
+func (l *loader) FetchClaims(ctx context.Context) ([]Claim, error) {
+	// Get the current claim count.
+	claimCount, err := l.claimFetcher.ClaimDataLen(&bind.CallOpts{})
+	if err != nil {
+		return []Claim{}, err
```
Comment (ajsutton) [inline op-challenger/fault/loader.go:76]: nit:
```suggestion
		return nil, err
```

File: op-challenger/fault/loader.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+// PushClaims pushes a list of claims into the [Game] state.
+func (l *loader) PushClaims(ctx context.Context, claims []Claim) error {
+	for _, claim := range claims {
+		if err := l.state.Put(claim); err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
Comment (ajsutton) [inline op-challenger/fault/loader.go:100]: Why does this method exist in the claim fetcher? Seems out of scope for it.

File: op-challenger/fault/loader.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+func NewLoader(log log.Logger, state Game, claimFetcher ClaimFetcher) *loader {
+	return &loader{
+		log:          log,
+		state:        state,
+		claimFetcher: claimFetcher,
+	}
+}
+
+// fetchClaim fetches a single [Claim] with a hydrated parent.
+func (l *loader) fetchClaim(ctx context.Context, arrIndex uint64) (*Claim, error) {
```
Comment (ajsutton) [inline op-challenger/fault/loader.go:43]: `ctx` isn't actually being used so can be removed from the API and from `FetchClaims`.

I'd probably modify it to just return `Claim` rather than a pointer as well given the only usage immediately dereferences it.  Just means returning `Claim{}, err` in a few places but that's fine.

File: op-challenger/fault/loader_test.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+// TestLoader_FetchClaims_Succeeds tests [loader.FetchClaims].
+func TestLoader_FetchClaims_Succeeds(t *testing.T) {
+	log := testlog.Logger(t, log.LvlError)
+	mockClaimFetcher := &mockClaimFetcher{}
+	loader := NewLoader(log, &mockGameState{}, mockClaimFetcher)
+	claims, err := loader.FetchClaims(context.Background())
+	require.NoError(t, err)
+	require.Equal(t, len(claims), 1)
+}
```
Comment (ajsutton) [inline op-challenger/fault/loader_test.go:126]: This isn't really testing all that much.  I'd say we want to be able to create `mockClaimFetcher` with a specified list of claims that it will return and then we should be able to assert that `loader.FetchClaims` returns that same list of claims.  Currently this test would pass even if `FetchClaims` was just:
```
		claim, _ := l.fetchClaim(ctx, 0)
		return []Claim{*claim}, nil
```
Admittedly you'd need to handle the error to make the next test pass too but there's nothing checking that it's actually retrieving the right number of claims.

**Review by ajsutton — COMMENTED** [review]
Left some comments, but I think the big issue here is that this code isn't reorg safe. It's quite possible that when `ClaimDataLen` is called there are 6 claims but a reorg then happens and there are then only 5 claims when `ClaimData` is called or that the order of claims in the array changes etc.

Each fetch needs to be given a specific block hash and use that in all calls to the contract to ensure it has a consistent view.  The code that calls this would likely just get the block hash of `latest` or possibly a block close to head but with a small follow distance and pass that it, but the fetcher shouldn't be using `latest` or it will get an inconsistent view.

File: op-challenger/fault/loader.go (hunk 5 above)
Comment (trianglesphere) [reply op-challenger/fault/loader.go:43]: Because this code does network calls, I think we should add the `ctx` to `CallOpts` (which is how it is uses in this API) instead of removing it.

File: op-challenger/fault/loader.go (hunk 3 above)
Comment (trianglesphere) [reply op-challenger/fault/loader.go:76]: also unfortunately go sometimes but not always treats `nil` differently from `[]Claim{}` depending on the circumstance.

Comment (ajsutton) [reply op-challenger/fault/loader.go:76]: True, but given we're returning an error there shouldn't be any assumptions made about the value returned here and `nil` is the more typical pattern at least from what I've seen.

File: op-challenger/fault/loader.go (hunk 4 above)
Comment (trianglesphere) [reply op-challenger/fault/loader.go:100]: maybe this could get moved to the game agent & the game agent takes a `FetchAllClaims` interface & then puts all of the claims into the game state

Comment (ajsutton) [reply op-challenger/fault/loader.go:100]: Yeah agreed - either in the game agent or make this a `AddAll` method on game state.  I lean towards the latter because that feels like a nicer API for the game state to expose and minimises the glue code required.

**Comment by refcell** [comment]
> Left some comments, but I think the big issue here is that this code isn't reorg safe. It's quite possible that when `ClaimDataLen` is called there are 6 claims but a reorg then happens and there are then only 5 claims when `ClaimData` is called or that the order of claims in the array changes etc.
> 
> Each fetch needs to be given a specific block hash and use that in all calls to the contract to ensure it has a consistent view. The code that calls this would likely just get the block hash of `latest` or possibly a block close to head but with a small follow distance and pass that it, but the fetcher shouldn't be using `latest` or it will get an inconsistent view.

Good callout - pinned claim fetching to a given `*big.Int` block number as that's what the CallOpts accepts.

I will leave the "latest" block number fetching out of the scope of this PR and up to the downstream usage of the claim loader to provide.

File: op-challenger/fault/loader.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+		state:        state,
+		claimFetcher: claimFetcher,
+	}
+}
+
+// fetchClaim fetches a single [Claim] with a hydrated parent.
+func (l *loader) fetchClaim(ctx context.Context, arrIndex uint64, blockNum *big.Int) (*Claim, error) {
+	fetchedClaim, err := l.claimFetcher.ClaimData(&bind.CallOpts{
+		Context:     ctx,
+		BlockNumber: blockNum,
```
Comment (trianglesphere) [inline op-challenger/fault/loader.go:51]: don't use `BlockNumber` here because we don't want to accidentally require archive state.

File: op-challenger/fault/loader.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+			Value:    fetchedClaim.Claim,
+			Position: NewPositionFromGIndex(fetchedClaim.Position.Uint64()),
+		},
+	}
+
+	if !claim.IsRootPosition() {
+		parentIndex := uint64(fetchedClaim.ParentIndex)
+		parentClaim, err := l.claimFetcher.ClaimData(&bind.CallOpts{
+			Context:     ctx,
+			BlockNumber: blockNum,
```
Comment (trianglesphere) [inline op-challenger/fault/loader.go:68]: ditto

File: op-challenger/fault/loader.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+	return &loader{
+		log:          log,
+		state:        state,
+		claimFetcher: claimFetcher,
+	}
+}
+
+// fetchClaim fetches a single [Claim] with a hydrated parent.
+func (l *loader) fetchClaim(ctx context.Context, arrIndex uint64, blockNum *big.Int) (*Claim, error) {
+	fetchedClaim, err := l.claimFetcher.ClaimData(&bind.CallOpts{
```
Comment (trianglesphere) [inline op-challenger/fault/loader.go:49]: can share `bind.CallOpts`

File: op-challenger/fault/loader.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+func NewLoader(log log.Logger, state Game, claimFetcher ClaimFetcher) *loader {
+	return &loader{
+		log:          log,
+		state:        state,
+		claimFetcher: claimFetcher,
+	}
+}
+
+// fetchClaim fetches a single [Claim] with a hydrated parent.
+func (l *loader) fetchClaim(ctx context.Context, arrIndex uint64, blockNum *big.Int) (*Claim, error) {
```
Comment (trianglesphere) [inline op-challenger/fault/loader.go:48]: generally returning a `Claim` is easier than returning `*Claim`

File: op-challenger/fault/loader.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+	return &claim, nil
+}
+
+// FetchClaims fetches all claims from the fault dispute game.
+func (l *loader) FetchClaims(ctx context.Context, blockNum *big.Int) ([]Claim, error) {
+	// Get the current claim count.
+	claimCount, err := l.claimFetcher.ClaimDataLen(&bind.CallOpts{
+		Context:     ctx,
+		BlockNumber: blockNum,
```
Comment (trianglesphere) [inline op-challenger/fault/loader.go:87]: ditto

**Comment by refcell** [comment]
Removing the `blockNumber` pin from callopts right now and tracking reorg hardening in a [new linear ticket](https://linear.app/optimism/issue/CLI-4164/harden-claim-loading).
Currently, we're working towards the MVP op-challenger/game solver and this will be a later task.

**Review by trianglesphere — APPROVED** [review]
lgtm

**Comment by ajsutton** [comment]
> Good callout - pinned claim fetching to a given `*big.Int` block number as that's what the CallOpts accepts.
> 
> I will leave the "latest" block number fetching out of the scope of this PR and up to the downstream usage of the claim loader to provide.

Pinning to a block number is not sufficient to avoid reorg issues unfortunately. It has to be a block hash.
